internal/controller/pod_controller.go

```diff
@@ -168,3 +189,136 @@ func (a *PodAdmissionWebhook) listOwners(ctx context.Context, object client.Obje
 	}
 	return owners, nil
 }
+
+// podEventHandlerRunner starts a podEventHandler when the manager
+// begins, ensuring that it only runs on the leader operator pod.
+type podEventHandlerRunner struct {
+	mgr manager.Manager
+	u   *workload.Updater
+	l   logr.Logger
+}
+
+// newPodEventHandlerRunner creates a new runner which will instantiate a
+// PodEventHandler during controller-runtime manager startup.
+func newPodEventHandlerRunner(mgr manager.Manager, u *workload.Updater, l logr.Logger) *podEventHandlerRunner {
+	return &podEventHandlerRunner{
+		mgr: mgr,
+		u:   u,
+		l:   l,
+	}
+}
+
+// newPodEventHandler creates a new PodEventHandler which implements
+func newPodEventHandler(ctx context.Context, c client.Client, u *workload.Updater, l logr.Logger) *podEventHandler {
+	return &podEventHandler{
+		ctx: ctx,
+		c:   c,
+		u:   u,
+		l:   l,
+	}
+}
+
+// NeedLeaderElection implements manager.LeaderElectionRunnable so that
+// the podEventHandler only runs on the leader, not on other redundant
+// operator pods.
+func (r *podEventHandlerRunner) NeedLeaderElection() bool {
+	return true
+}
+
+// Start implements manager.Runnable which will start the informer receiving
+// pod change events on the operator's leader instance.
+func (r *podEventHandlerRunner) Start(ctx context.Context) error {
+	h := newPodEventHandler(ctx, r.mgr.GetClient(), r.u, r.l)
+
+	i, err := r.mgr.GetCache().GetInformerForKind(ctx, schema.GroupVersionKind{
+		Group:   "",
+		Version: "v1",
+		Kind:    "Pod",
+	})
+	if err != nil {
+		return fmt.Errorf("Unable to get pod informer, %v", err)
+	}
+
+	_, err = i.AddEventHandler(h)
+	if err != nil {
+		return fmt.Errorf("Unable to register pod event handler, %v", err)
+	}
+	return nil
+
+}
+
+// podEventHandler receives pod changed events for all pods in the K8s cluster
+// to ensure that any pods that should have proxy sidecars are configured
+// correctly. If a pod is not configured correctly, and is not running, then
+// the operator will delete the pod.
+type podEventHandler struct {
+	ctx context.Context
+	c   client.Client
+	u   *workload.Updater
+	l   logr.Logger
+}
+
+// OnAdd implements cache.ResourceEventHandler, gets called by the cache
+// informer when a Pod is added to the k8s cluster.
+func (h *podEventHandler) OnAdd(obj interface{}) {
```

Comment:
Does OnAdd() events also trigger when the mutating webhook triggers? I'm wondering if pod objects might be processed twice.

Reply:
Discussed offline. OnAdd will preform a no-op reprocessing after the mutating webhook modifies a pod.